Add PlanLogView

Vielheim opened the pull request:
PlanLogView shows all the Comments, Upvoted Users (summarised) and Plan changes (summarised) inside one ScrollView. It replaces the CommentsView inside the PlanDisplayView. 
- Still looking to improve code reuse for displaying Comments
- Looking to reduce the logic inside PlanLogView. 

Some changes to declaration:
- PlanView now holds the view models as `@StateObjects`
- Children (PlanDisplayView and its components) now observe the view models
- AddCommentVM is now a delegate (changes to plan version in CommentsVM is propagated to the AddCommentVM)

I did it this way so that the `PlanUpvoteView` and `PlanLogView` can share the fetch upvotes logic. Then I did the same for Comments just to be consistent.

Commit:
Add PlanLogView

- It will display all comments, updates to Plans and upvotes (for each version)
- Because the viewmodels are shared within PlanDisplayView, i set the viewmodels to be @StateObject in PlanView, @ObservedObject in PlanDisplayView and its children.
This allows us to maintain the same state across the children. Do advise if there is a better way to declare them.

- AddCommentViewModel is now attached to CommentsViewModel (part of the PlanEventDelegate chain) so when the plan version updates, the AddComment also gets updated.

Commit:
Add PlanModifierMap to PlanDisplayViewModel

keanecjy commented on TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanDisplayView.swift:
```diff
@@ -10,8 +10,8 @@ import SwiftUI
     @ObservedObject var planDisplayViewModel: PlanDisplayViewModel<T>
-    let commentsViewModel: CommentsViewModel
-    let planUpvoteViewModel: PlanUpvoteViewModel
+    @ObservedObject var commentsViewModel: CommentsViewModel
+    @ObservedObject var planUpvoteViewModel: PlanUpvoteViewModel
```
was it wrong previously?

Vielheim replied:
I'm not very sure also but I don't think so? Previously these VMs don't need to be stateful/observed in the views until they reach the components when they're set to @StateObject. But now multiple components are sharing the views so I shift the `@StateObject` to the parent view.

keanecjy replied:
Ah I see

Commit:
shift view generation into ViewFactory

Commit:
update published in PlanDisplayVM

keanecjy commented on TourMate/TourMate/Frontend/Common/ViewFactory/ViewFactory+PlanLogView.swift:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    func getComments(commentsViewModel: CommentsViewModel, version: Int) -> some View {
```
Are we no longer using commentsview?

Vielheim replied:
I'm trying to see how to integrate the PlanLogView and CommentsView together. It didn't make sense at first to pass like `planDisplayViewModel` and `planUpvoteViewModel` into a CommentsView because it should be mainly for displaying comments.

keanecjy replied:
Should we put this in their own view instead of in the view factory?

Vielheim replied:
I will abstract the View part into a Component but I think I might keep the logic (like generating the Strings to display) inside the view factory

keanecjy commented:
The upvotes does not show in the plansview anymore

keanecjy commented on TourMate/TourMate/Frontend/Plan/Views/PlanUpvoteView/PlanUpvoteView.swift:
```diff
@@ -9,11 +9,11 @@ import SwiftUI
 struct PlanUpvoteView: View {
 
-    @StateObject var viewModel: PlanUpvoteViewModel
+    @ObservedObject var viewModel: PlanUpvoteViewModel
```
i think all views using this would have to change to stateobject before passing creating this

Vielheim replied:
Ok I updated it

Commit:
add ActionableContentView and ScrollableContentView

- They abstract some of the UI logic common in the PlanView (comments and Plan log)

Commit:
Update declaration of PlanUpvoteViewModel to @StateObject

Commit:
extract components, make CommentListView reusable

keanecjy commented on TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanDisplayViewModel.swift:
```diff
@@ -14,12 +14,18 @@ class PlanDisplayViewModel<T: Plan>: ObservableObject {
     @Published var planLastModifier: User
 
     var allVersionedPlans: [T]
+    @Published var planModifierMap: [Int: User] // version to user map
```
where is this plan modifier map going to be used?

keanecjy reviewed (approved):
Looks good
